### DISCUSSION TITLE
Let the user know once the connection is successful.

### DISCUSSION THREAD

**Discussion by Seluj78** [shellhub-io/shellhub#3476] [Ideas]
When connecting through the command line, using

`ssh -p PORT <user>@<hostname>@<domain>`

We have no clue as to when the connection is successful, especially if you're using the same shell locally and remotely

When the connection is successful, we just get dropped inside the new shell on the machine.

I would love a little configurable message to be displayed upon connection :)

**Comment by gustavosbarreto** [comment]
We are considering implementing a feature to display a message, following the standard behavior of OpenSSH, which typically outputs the system's default message of the day (motd) located at `/etc/motd`. We'll keep you posted on any updates regarding this enhancement.

Appreciate your feedback!

**Reply by Seluj78** [reply]
Great thank you very much. It would be very useful to us to be able to display `Do not turn off the docker ssh agent` because I am certain some users will turn it off and lock themselves out of the machine (Although i haven't tried what happened when SSHed into a machine through shellhub and you stop the agent)

**Reply by Seluj78** [reply]
Ideally we could choose to use `/etc/motd` or use a custom message that we can edit from the UI/API ?

**Reply by heiytor** [reply]
Hi, we think it's a really good idea, but we need more context about what you want with this feature.

Currently, it's already possible by setting a text in the [`/etc/motd`](https://en.wikipedia.org/wiki/Message_of_the_day). However, we can implement a similar feature but at *namespace* level.

Additionally, we would like to understand your preference regarding the display of both the *motd* (at the "device" level) and the message at the namespace level, or if you would prefer only one of them to be displayed. For example, *motd* can be configured to overwrite the namespace message.

**Reply by Seluj78** [reply]
Just for info, because I never tried before, I just stopped the agent while ssh'ed through shellhub and the connection did close so it is important for us to warn the user. I fixed the machine by just ssh'ing from AWS and running the `curl` command again but with our users it might be a pain to do every so often.

**Reply by Seluj78** [reply]
As for our usecase:

We provide small EC2 instances for our users to do stuff in. We have over 2500 instances that need to be accessed infrequently through SSH for those users, either via a webshell or directly from their machine.

We used to use elastic IPs but AWS is changing their pricing + we are at the limit of 500 elastic IPs, so it's getting complicated for us to manage that and it's why we're switching to shellhub.

Our use case is: User needs access to their machine. They start it, it becomes available and they can SSH through our webshell or from their machine using the shellhub tunnel. (UI will be only for us admins/devs). Some users need to use docker on their machines and I am afraid that they might do a `docker stop` or something bad like that and their connection will close, and it's a new support ticket on our end to handle. It's an easy fix, but time consuming.

That's why I need some kind of MOTD to show up. Something like `Please do not stop the shellhub/agent docker container or you will be locked out of your machine`.
Ideally it would be customizable from the shellhub UI and API so we can change it easily. For us, modifying the `/etc/motd` of every single 2500+ machines is not great, even with terraform :)

Let me know if you need more info

**Reply by Seluj78** [reply]
> Currently, it's already possible by setting a text in the [/etc/motd](https://en.wikipedia.org/wiki/Message_of_the_day). However, we can implement a similar feature but at namespace level.

I don't think it works, or at least not on my end.


Connecting directly to AWS
```
➜  pemkeys ssh -i "keypair.pem" ubuntu@xxx.eu-west-1.compute.amazonaws.com
Welcome to Ubuntu 22.04.3 LTS (GNU/Linux 6.2.0-1018-aws x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/pro

  System information as of Wed Jan 31 20:04:57 UTC 2024

  System load:  0.0                Processes:                100
  Usage of /:   18.2% of 15.32GB   Users logged in:          0
  Memory usage: 27%                IPv4 address for docker0: 172.17.0.1
  Swap usage:   0%                 IPv4 address for eth0:    172.31.42.245

 * Ubuntu Pro delivers the most comprehensive open source security and
   compliance features.

   https://ubuntu.com/aws/pro

Expanded Security Maintenance for Applications is not enabled.

15 updates can be applied immediately.
2 of these updates are standard security updates.
To see these additional updates run: apt list --upgradable

Enable ESM Apps to receive additional future security updates.
See https://ubuntu.com/esm or run: sudo pro status


Last login: Wed Jan 31 19:51:30 2024 from xxx
➜  ~
```

From shellhub:
```
➜  .ssh ssh -p 2299 -i id_ed25519 ubuntu@pre-production.test@shellhub.example.com
➜  ~
```

Both are successful, but shellhub currently doesn't display the MOTD

**Comment by heiytor** [comment]
Hi, @Seluj78 

I've spent a couple of minutes working on a prototype of this feature, which you can find in #3508. The branch is definitely broken, with many TODOs that I still need to address, and it's not yet available on the UI. However, for testing purposes, it should work fine.

To test the feature, you can make a PUT request to `/api/namespaces/:tenant/connection_announcement` with the following JSON body:
```json
{
	"announcement": "your message here"
}
```
To verify the announcement, you can make a simple GET request to `/api/namespaces/:tenant`. The announcement should be in `settings.connection_announcement`.

With a proper connection_announcement set, when connecting to a device within this namespace, the message should appear if the connection is successful:
![image](https://github.com/shellhub-io/shellhub/assets/107213601/8320c253-35c7-4e09-99ee-b552a6f45cc0)

I'm not sure when this feature will be included in a release.

**Reply by Seluj78** [reply]
Wow this is amazing! I'll try it out when I can !

I don't have any remarks seeing what you showed me, it's exactly what I would need for now, although I can't make any promises I won't ask for more in the future 👀

**Reply by Seluj78** [reply]
Maybe we could have announcements with tags or namespaces ? So we can have different announcements based on machine types (for our usecase)

**Reply by heiytor** [reply]
> I don't have any remarks seeing what you showed me, it's exactly what I would need for now, although I can't make any promises I won't ask for more in the future 👀

We will definitely be happy to receive your questions and suggestions. :)

> Maybe we could have announcements with tags or namespaces ? So we can have different announcements based on machine types (for our usecase)

Creating this feature for the namespace is quite simple. However, implementing it for tags might be a bit more complex due to the way tags are handled in the backend. We will investigate this possibility, and I'll keep you updated.

**Reply by Seluj78** [reply]
> Creating this feature for the namespace is quite simple. However, implementing it for tags might be a bit more complex due to the way tags are handled in the backend. We will investigate this possibility, and I'll keep you updated.

Thank you !

My idea behind it, so you understand my use case (it might not be useful for everyone using your product) is that we have a few different types of machines we give to our users, and we would like to be able to display a custom message based on what type of machine they are using. It's definetly doable with namespaces, and I think it is what I will use when implementing shellhub in my backend so I can sort through machines easily, but tags would be cool as well.

**Comment by heiytor** [comment]
Hi, @Seluj78,

The feature has already been merged, and you can check the final version on the master branch.
